### PR TITLE
feat: mvp

### PR DESCRIPTION
**Pull request by Yohe-Am**

### CODE REVIEW

File: cli/core/hooks.ts [hunk 1]
```diff
@@ -8,10 +8,56 @@ const log = console.log;
...
-const module = await import(Deno.args[0]);
-console.log = log;
-module.ghjk.runCli(Deno.args.slice(1));
+const mod = await import(Deno.args[0]);
+mod.ghjk.runCli(Deno.args.slice(1));
     `,
+  "hooks/hook.sh": `
+ghjk_already_run=false
+
+ghjk_hook() {
```
Comment (zifeo) [inline cli/core/hooks.ts:17]: https://github.com/rcaloras/bash-preexec might be useful to support all bash issues, we vendored with lade

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -73,12 +73,12 @@ repos:
           - ts
           - tsx
         files: ^website/
-      - id: version
-        name: "Lock versions"
-        always_run: true
-        language: system
-        entry: bash -c 'deno run -A dev/lock.ts --check'
-        pass_filenames: false
+      # - id: version
```
Comment (zifeo) [inline .pre-commit-config.yaml:76]: you can remove everything that does not make sense, I forgot to adapt the file

File: cli/hooks.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  log("[ghjk.ts]", ...args);
+};
+const mod = await import(Deno.args[0]);
+console.log = log;
+mod.ghjk.runCli(Deno.args.slice(1), mod.options);
+    `,
+
+  // the hook run before every prompt draw in bash
+  "hooks/hook.sh": `
+ghjk_already_run=false
```
Comment (zifeo) [inline cli/hooks.ts:19]: can you explain when this flag is required?

File: cli/hooks.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        eval $GHJK_CLEANUP
+        unset GHJK_CLEANUP
+    fi
+    cur_dir=$PWD
+    while [ "$cur_dir" != "/" ]; do
+        if [ -e "$cur_dir/ghjk.ts" ]; then
+            envDir="$HOME/.local/share/ghjk/envs/$(echo "$cur_dir" | tr '/' '.')"
+            if [ -d "$envDir" ]; then
+                PATH="$envDir/shims:$(echo "$PATH" | tr ':' '\n' | grep -vE "^$HOME/\.local/share/ghjk/envs" | tr '\n' ':')"
+                PATH="$\{PATH%:\}"
```
Comment (zifeo) [inline cli/hooks.ts:37]: not sure what this does exactly?

**Review by zifeo — COMMENTED** [review]
very nice, can you add GitHub action? at one points we will be able to use ghjk for its own tests

File: cli/hooks.ts (hunk 3 above)
Comment (Yohe-Am) [reply cli/hooks.ts:20]: Ah, that should no longer be necessary once I bring in https://github.com/rcaloras/bash-preexec into use.

File: cli/hooks.ts (hunk 4 above)
Comment (Yohe-Am) [reply cli/hooks.ts:38]: That's doing the fish equivalent of `set --global PATH $envDir/shims $(string match --invert --regex "^$HOME\/\.local\/share\/ghjk\/envs" $PATH)`.

Which is to say that it's adding the shim dir to path. It removes it first if it already exists to avoid duplicates.

Comment (Yohe-Am) [reply cli/hooks.ts:38]: Bash exposes the `$PATH` entries in a string separated with colons. The code changes the string to use `\n` separators instead before it uses `grep` to remove the old shim. It then changes it back to a colon separated string before setting it.

File: cli/hooks.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+  await unpackVFS(shareDir);
+  const shell = await detectShell();
+  if (shell === "fish") {
+    await filterAddFile(
+      std_path.resolve(homeDir, ".config/fish/config.fish"),
+      /\.local\/share\/ghjk\/hooks\/hook.fish/,
+      "source $HOME/.local/share/ghjk/hooks/hook.fish",
+    );
+  } else if (shell === "bash") {
+    await filterAddFile(
```
Comment (afmika) [inline cli/hooks.ts:234]: Support for zsh might also be great  for the future.

Comment (afmika) [reply cli/hooks.ts:234]: Also maybe use endsWith or includes? `detectShell` seems to return the path.

**Review by afmika — APPROVED** [review]
Overall great!

**Review by zifeo — APPROVED** [review]
Wow,, amazing job! Little questions/suggestions

File: README.md [hunk 6]
```diff
@@ -66,10 +66,21 @@ and looks as follows (abstracting away some implementation details):
...
+- [x] wasm-tools
+- [x] cargo-insta
 - hash verifiable dependencies (timestamp)
+- hide the `Deno` object in an abstraction
+- support windows
+- [ ] installation tools
+  - [ ] untar
+  - [ ] xz
+  - [ ] git
+- [ ] Find a way to make copy ops atomic
```
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply README.md:86]: Right now, sans lockfile, the way we're checking if plugins is installed is by checking if the file exists. The plugins download/install into a `tempDir` first then copy it to the actual dir when they're done. The `asdf` guidelines also discourage adding anything to the install dir if it's not successful. But if the users `kill -9`s right in the middle of a copy, things break. 

Ticking that off will mainly be having the `tmpDir`s within the same filesystem and using `mv`. `mv` is atomic but the default `tmpDir` is usually located in a separate `tmpfs`.

File: cli/sync.ts [hunk 7]
```diff
@@ -1,85 +1,447 @@
...
+      const depInstallId = getInstallId(depInstall);
+      const depArtifacts = artifacts.get(depInstallId);
+      if (!depArtifacts) {
+        throw new Error(
+          `artifacts not found for plug dep "${depInstallId}" when installing "${installId}"`,
+        );
+      }
+      const depShimDir = std_path.resolve(depShimsRootPath, depInstallId);
+      await Deno.mkdir(depShimDir);
+      // TODO: expose LD_LIBRARY from deps
```
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply cli/sync.ts:124]: Yeah, we're providing plugins their dependencies by creating a special temporary shim directory for each invocation. This should allow us to cleanly set the `--allow-run=/to/shim` we need. What's missing is providing shims for their dep libraries as well.

File: cli/sync.ts [hunk 8]
```diff
@@ -1,85 +1,447 @@
...
+      if (conflict) {
+        throw new Error(
+          `duplicate env var found ${key} from installs ${instId} & ${conflict[1]
+          }`,
         );
+      }
+      env[key] = [val, instId];
+    }
+  }
+  // FIXME: prevent malicious env manipulations
```
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply cli/sync.ts:210]: I was mainly thinking of a blacklist for things like PATH.

File: cli/sync.ts [hunk 9]
```diff
@@ -1,85 +1,447 @@
...
+      const glob = file.startsWith("/")
+        ? file
+        : std_path.joinGlobs([installPath, file], { extended: true });
+      for await (const entry of std_fs.expandGlob(glob)) {
+        foundTargetPaths.push(entry.path);
+      }
+      continue;
+    }
+    const filePath = std_path.resolve(installPath, file);
+    const fileName = std_path.basename(filePath); // TODO: aliases
```
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply cli/sync.ts:323]: Yeah, I'm thinking a field on `InstallConfig` to rename shims of specific bins directly `ghjk.ts`.

File: core/utils.ts [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+export type SpawnOptions = {
+  cwd?: string;
+  env?: Record<string, string>;
+  pipeInput?: string;
+  // pipeOut?: WritableStream<Uint8Array>;
+  // pipeErr?: WritableStream<Uint8Array>;
+};
+
+// FIXME: pita function responsible for test flakiness
```
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply core/utils.ts:31]: Yeah, it sometimes breaks because of closed pipes. Also, minor behavioral differences when running within workers.

[the remark this answers is not in the document]
Comment (Yohe-Am) [reply core/utils.ts:31]: I'll go looking for a library for this. Something more ergonomic to expose to the plugs might be nice.

File: README.md [hunk 11]
```diff
@@ -66,10 +66,21 @@ and looks as follows (abstracting away some implementation details):
...
-- multiple version of the same package (e.g. rust stable and rust nightly)
-- wasmedge
-- python with virtual env dir
-- poetry
-- pnpm
-- mold({ if: Deno.build.os === "Macos" })
+- multiple version of the same package (e.g. rust stable and rust nighted)
+- [x] wasmedge
+- [x] jco
+- [ ] python with virtual env dir
```
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply README.md:72]: We're able to install python and poetry fine through asdf compat but it's not ideal. 

The first think thing we'll have to tackle is lockfiles.  [`asdf-python`](https://github.com/asdf-community/asdf-python) relies directly on [`pyenv`](https://github.com/pyenv/pyenv) which builds [`stackless python`](https://github.com/stackless-dev/stackless/) from source.  As it stands, everything is installed from scratch on every `ghjk sync` and it'd be un-ideal that every plugin that uses `pip`, `pre-commit` for example, had to first build `python` not to mention install deps like `cc` and `perl` required for that.  With a lockfiles, at least we'll only need to build once. Ideally, even share some of those across environments.

The second issue is that when doing `pip install foo`, it installs it to the configured virtual environment. Ideally, we'll have install to `$ghjkEnv/installs/foo/v1.2.3/` and I intend to try and hack it there but might not be possible.  Which begs the question, should the virtual environment be located within the ghjk environment or CWD? If we go with the former, plugins will need to have a `working` dir besides the `install` and `download` dirs that currently exist. 

Another option is to afford a special case for `pip installs` like we currently do `asdf` which has a special working dir in `$envDir/asdf`. But special cases are usually a symptom of something wrong with an abstraction so maybe we can do better.

[the remark this answers is not in the document]
Comment (zifeo) [reply README.md:72]: > everything is installed from scratch on every ghjk sync
if on the CI, the .local/share/ghjk folder should be cached, if locally, if the shim already exist in .local/share/ghjk/installs I would not rebuild it. By default I would only install (and compile for python) if there is no match installs. Maybe we will need `ghjk sync --fresh` or more fine-grained control. Do you see other possible scenario?

>should the virtual environment be located within the ghjk environment or CWD?
I would try to get it stored on .local/share/ghjk/shims/a-b-c/python/name-of-venv and manipulate the env to get the bindings right

**Comment by Yohe-Am** [comment]
Some comments on the refactor:
- I've renamed Plugs to Ports to avoid the generic name.
- Ghjk is now made up of Modules like the Ports module and the upcoming Task module. 
    -  Modules can expose cli subcomands
         - `ghjk sync` is now `ghjk ports sync`. Abstractions for shorthands should be possible.

File: install/hooks/fish.fish [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+                echo $envDir
+                set_color normal
+            end
+            set ghjk_alias "deno run --unstable-worker-options -A $HOME/.local/share/ghjk/hooks/entrypoint.ts $cur_dir/ghjk.ts"
+            return
+        end
+        set cur_dir (dirname $cur_dir)
+    end
+    clean_up_paths
+    set ghjk_alias "echo 'No ghjk.ts config found.'"
```
Comment (zifeo) [inline install/hooks/fish.fish:45]: to keep this file as simple as possible, I would call entrypoint.ts to generate the errors ourselves

File: install/hooks/fish.fish [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+end
+
+# the alias value could be changed by init_ghjk
+# to execute the appropriate cmd based on ghjk.ts
+set ghjk_alias "echo 'No ghjk.ts config found.'"
+function ghjk
+    eval $ghjk_alias $argv
```
Comment (zifeo) [inline install/hooks/fish.fish:52]: why is that needed?

File: install/mod.ts [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  "hooks/bash-preexec.sh": await importRaw(
+    "https://raw.githubusercontent.com/rcaloras/bash-preexec/0.5.0/bash-preexec.sh",
+  ),
+
+  "hooks/.zshenv": (
+    await importRaw(import.meta.resolve("./hooks/zsh.zsh"))
+  )
+    .replaceAll("__var_LD_LIBRARY_ENV__", LD_LIBRARY_ENV),
+  // the hook run before every prompt draw in bash
+  "hooks/hook.sh": (
```
Comment (zifeo) [inline install/mod.ts:35]: any reason for .sh instead of the .bash?

File: mod.ts [hunk 15]
```diff
@@ -1,2 +1,60 @@
-// this is only a shortcut for the cli
-export * from "./cli/mod.ts";
+//! This module is intended to be re-exported by `ghjk.ts` config scripts. Please
+//! avoid importing elsewhere at it has side-effects.
```
Comment (zifeo) [inline mod.ts:2]: do you see an enforceable strategy to avoid such issues?

File: tests/test.Dockerfile [hunk 16]
```diff
@@ -0,0 +1,42 @@
+FROM docker.io/denoland/deno:debian-1.38.0
+
+RUN set -eux; \
+    export DEBIAN_FRONTEND=noninteractive; \
+    apt update; \
+    apt install --yes \
+    # test deps
+    fish zsh \
```
Comment (zifeo) [inline tests/test.Dockerfile:8]: can we fix the version? or maybe use the docker image from the hub and copy the deno bin inside?

File: mod.ts (hunk 15 above)
Comment (Yohe-Am) [reply mod.ts:2]: Actually, the side-effects are no longer necessary for `mod.ts` with the new approach but yeah, there are some global variables that we interact that are only available in certain context and some principled approach will help. 

Right now, I'm thinking a locked down global variable describing what context the module's been loaded in that code interacts with global state can assert against before continuing.

File: install/mod.ts (hunk 14 above)
Comment (Yohe-Am) [reply install/mod.ts:35]: I was hoping to get the file POSIX compliant but it's an uphill battle. Will rename.

File: install/hooks/fish.fish (hunk 13 above)
Comment (Yohe-Am) [reply install/hooks/fish.fish:52]: Premature abstraction in case we needed to do more than what fits in an alias but yeah, best to simplify all things. I'll go back to the pure alias approach.

File: LICENSE.md [hunk 17]
```diff
@@ -1,28 +1,23 @@
-Mozilla Public License Version 2.0
-==================================
+# Mozilla Public License Version 2.0
```
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply LICENSE.md:1]: I did not change the license myself actually. Must have been a pre-commit hook, I'll revert those changes.

File: README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply README.md:72]: Changed to issue at #3 .

File: cli/sync.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Yohe-Am) [reply cli/sync.ts]: Best implement this after we've found a usecase.

**Comment by Yohe-Am** [comment]
Comments on the latest changes:

- ZSH support is in and it's piggybacking of the mostly POSIX compliant support for bash. `precmd` is natively supported in ZSH so it was straightfwd.
- The exported functions like `ghjk` and `init_ghjk` were previously there for non-interactive support as aliases are not recognized there. Exported functions were not not great either as only `bash` and `fish` properly support them and things like [Shellshock](https://en.wikipedia.org/wiki/Shellshock_(software_bug)) means they're generally frowned upon. We're now using installing an executable `ghjk` into the users PATH that simply calls `deno run --unstable -A ghjkRepo/main.ts`. This less self-contained than before but it works well everyhwere.

File: install/hooks/bash.sh [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+# but even reliably resolving it's address
+# requires bash extensions. 
+if [ -n "${BASH_SOURCE+x}" ]; then
+    hooksDir=$(dirname -- "$(readlink -f -- "${BASH_SOURCE}")")
+    . "$hooksDir/bash-preexec.sh"
+fi
+
+# use precmd to check for ghjk.ts before every prompt draw
+# precmd is avail natively for zsh
+precmd() {
```
Comment (zifeo) [inline install/hooks/bash.sh:49]: should we test for cd, pushd, and popd to avoid init when the path has not changed?

**Review by zifeo — APPROVED** [review]
Neat!

File: install/hooks/bash.sh (hunk 18 above)
Comment (Yohe-Am) [reply install/hooks/bash.sh:49]: Nah, we still want to load the latest values from the loaders on every prompt as they will change after  `ghjk port sync` even though PWD hasn't changed.

Comment (zifeo) [reply install/hooks/bash.sh:49]: you could add ghjk to this list?

Comment (zifeo) [reply install/hooks/bash.sh:49]: the rational behind is that to keep the script shell native, it is usually < 1ms for 3 levels of directory (which is way faster than loading a binary to memory), but it can seriously increase to 100ms (= noticeable) if you are deep in your nested folders, thus running each time the init might feel too much (we would need to benchmark under what conditions we may need to switch to a binary)

**Review by zifeo — APPROVED** [review]

File: action.yml [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    # FIXME: find a way to get commit sha of current executing action version
+    # default: $GITHUB_SERVER_URL/$GITHUB_ACTION_REPOSITORY/raw/feat/
+    default: './install.ts'
+  sync:
+    description: 'Disable to skip syncing ports'
+    required: true
+    default: true
+runs:
+  using: "composite"
+  steps:
```
Comment (zifeo) [inline action.yml:16]: install deno if not present? + cache path?

File: install/hooks/bash.sh [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+# BLUE='\033[0;34m'
+ansi_nc='\033[0m' # No Color
+
+init_ghjk() {
+    if [ -n "${GHJK_CLEANUP+x}" ]; then
+        eval "$GHJK_CLEANUP"
+    fi
+    unset GHJK_CLEANUP
+    unset GHJK_LAST_LOADER_PATH
+    unset GHJK_LAST_LOADER_TS
```
Comment (zifeo) [inline install/hooks/bash.sh:18]: to be included in the cleanup?

File: install/hooks/bash.sh [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+                printf "${ansi_red}[ghjk] Uninstalled runtime found, please sync...${ansi_nc}\n"
+                printf "$envDir\n"
+            fi
+            return
+        fi
+        cur_dir="$(dirname "$cur_dir")"
+    done
+}
+
+# onlt load bash-prexec if we detect bash
```
Comment (zifeo) [inline install/hooks/bash.sh:42]: typo?

File: install/hooks/bash.sh [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+    . "$myDir/bash-preexec.sh"
+fi
+
+export LAST_PWD="$PWD"
+# use precmd to check for ghjk.ts before every prompt draw
+# precmd is avail natively for zsh
+precmd() {
+    if [ "$LAST_PWD" != "$PWD" ] || (
+        # if the last detected loader has been touched
+        [ -n "${GHJK_LAST_LOADER_PATH+x}" ] && [ $(stat -c "%Y" "$GHJK_LAST_LOADER_PATH" | tr -d '\n') != $(("$GHJK_LAST_LOADER_TS")) ]
```
Comment (zifeo) [inline install/hooks/bash.sh:57]: Running the loader on each change of directory sounds ok and avoid the complexity of checking the loader (speed should be ok unless the guy keeps changing dir :)

Comment (Yohe-Am) [reply install/hooks/bash.sh:57]: I considered loading directly on every prompt but I'm not sure how well I we can trust the env vars. They might have been set in another parent shell in a different pwd. The current approach instead just checks if the cached loader has been touched but then t tries to locate the loader from scratch in case things aren't as they should be.

File: install/hooks/bash.sh (hunk 20 above)
Comment (Yohe-Am) [reply install/hooks/bash.sh:18]: They're too hook specific for that I think. We might have use different env vars for different shells.

File: action.yml (hunk 19 above)
Comment (Yohe-Am) [reply action.yml:16]: I think a rewrite into a [Js based action](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action) is necessary to impl the caching. Will tackle that in that PR.

File: install/hooks/bash.sh (hunk 21 above)
Comment (Yohe-Am) [reply install/hooks/bash.sh:42]: Will fix. About to merge this and don't want to loose the approval :)
